lib/api/multiObjectDelete.js

```diff
@@ -242,114 +286,152 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                    isDeleteMarker = true;
+                    deleteMarkerVersionId = versionIdUtils.encode(
+                        versionId, config.versionIdEncodingType);
+                    // In this case we are putting a new object (i.e., the delete
+                    // marker), so we decrement the numOfObjectsRemoved value.
+                    numOfObjectsRemoved--;
+                    // If trying to delete a delete marker, DeleteMarkerVersionId equals
+                    // deleteMarker's versionID and DeleteMarker equals true
+                } else if (objMD && objMD.isDeleteMarker) {
+                    isDeleteMarker = true;
+                    deleteMarkerVersionId = entry.versionId;
                 }
-                const objLockInfo = new ObjectLockInfo({
-                    mode: objMD.retentionMode,
-                    date: objMD.retentionDate,
-                    legalHold: objMD.legalHold || false,
+                successfullyDeleted.push({
+                    entry, isDeleteMarker,
+                    deleteMarkerVersionId,
                 });
+                return moveOn();
+            });
+        },
+            // end of forEach func
+            err => {
+                // Batch delete all objects
+                const onDone = () => callback(err, quietSetting, errorResults, numOfObjectsRemoved,
+                    successfullyDeleted, totalContentLengthDeleted, bucket);
 
-                // If the object can not be deleted raise an error
-                if (!objLockInfo.canModifyObject(hasGovernanceBypass)) {
-                    log.debug('trying to delete locked object');
-                    return callback(objectLockedError);
+                if (err && deleteFromStorage.length === 0) {
+                    log.trace('no objects to delete from data backend');
+                    return onDone();
                 }
-
-                return callback(null, objMD, versionId);
-            },
-            (objMD, versionId, callback) => {
-                const options = preprocessingVersioningDelete(
-                    bucketName, bucket, objMD, versionId, config.nullVersionCompatMode);
-                const deleteInfo = {};
-                if (options && options.deleteData) {
-                    deleteInfo.deleted = true;
-                    if (objMD.uploadId) {
-                        // eslint-disable-next-line
-                        options.replayId = objMD.uploadId;
+                // If error but we have objects in the list, delete them to ensure
+                // consistent state.
+                log.trace('deleting objects from data backend');
+                const deleteLog = logger.newRequestLoggerFromSerializedUids(log.getSerializedUids());
+                return data.batchDelete(deleteFromStorage, null, null, deleteLog, err => {
+                    if (err) {
+                        log.error('error deleting objects from data backend', { error: err });
```

Comment:
when this happens, all the "metrics" computed above may be wrong : numOfObjectsRemoved, successfullyDeleted, totalContentLengthDeleted...

* if `batchDelete` works somewhat "atomically" (e.g. it succeeds to delete all OR fails and delete nothing), then the logic can be tweak to count properly;
* otherwise, we may need that function to return the "list" of objects actually delete and their size, so we can do the appropriate count

Reply:
The current behavior is that, in case of any error, the function returns, and the metrics are unchanged, as they are only computed based on (successful) operations on the metadata, not the storage. A failure in deleting the actual data would lead to the whole API to fail. The only difference I see here is the number or orphans we would get, more than an issue with the metrics. When the API fails, we do not push any metric... So to me, this is not something to handle, as we push no metrics anyway.

The main question is more about handling or orphans: this can be mitigated by pre-splitting the array from `1000` to `1000/${multiObjectDeleteConcurrency}` as this will be aligned with the current parallelism, without the drawback of waiting and doing the calls one by one. I will make the associated changes.